core/network/examples/peer_gossip.rs

```diff
@@ -0,0 +1,242 @@
+        let mut stream_guard = self.stream.try_lock().expect("aa");
+        stream_guard.write_all(&*request_bytes).await.unwrap();
+
+        loop {
+            // Lock the listener and wait for a new connection
+            let clone = self.listener.clone();
+            let listener = clone.try_lock().expect("Unable to lock listener");
+
+            match listener.accept().await {
+                Ok((mut stream, _)) => {
+                    let mut buf = [0u8; 1024];
+                    match stream.read(&mut buf).await {
+                        Ok(n) => {
+                            if n == 0 {
+                                break;
+                            }
+                            println!("Received a message of length: {}", n);
+                            on_response(buf[0..n].to_vec());
+                        }
+                        Err(e) => {
+                            eprintln!("Error reading from stream: {}", e);
+                            break;
+                        }
+                    }
+                    break;
+                }
+                Err(e) => {
+                    eprintln!("Error accepting connection: {}", e);
+                    break;
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone, Hash, Debug)]
+struct TestGossipableType {
+    pub id: Id,
+}
+
+impl Default for TestGossipableType {
+    fn default() -> Self {
+        TestGossipableType {
+            id: Default::default(),
+        }
+    }
+}
+
+impl Gossipable for TestGossipableType {
+    fn get_id(&self) -> Id {
+        self.id
+    }
+
+    fn serialize(&self) -> Result<Vec<u8>, Box<dyn Error>> {
+        Ok(self.id.to_vec())
+    }
+
+    fn deserialize(&mut self, bytes: &[u8]) -> Result<(), Box<dyn Error>> {
+        self.id = Id::from_slice(bytes);
+        Ok(())
+    }
+}
+
+// Mock implementation for the Set trait
+//ToDo Should we move all tests to a new file ?
+#[derive(Debug, Clone, Hash)]
+pub struct MockSet<TestGossipableType> {
+    pub set: Vec<TestGossipableType>,
+}
+
+impl<T> MockSet<T> {
+    pub fn len(&self) -> usize {
+        println!("{}", self.set.len());
+        self.set.len()
+    }
+}
+
+impl<T: Gossipable + Sync + Send + Clone + Hash> Set for MockSet<T> {
+    type Item = T;
+    fn add(&mut self, _gossipable: T) -> Result<(), Box<dyn Error>> {
+        self.set.push(_gossipable.clone());
+        Ok(())
+    }
+
+    fn iterate(&self, _f: &dyn FnMut(&T) -> bool) {
+        // Do nothing
+    }
+
+    fn fetch_elements(&self) -> Self::Item {
+        self.set.get(0).unwrap().clone()
+    }
+}
+
+async fn fake_handler_server_logic(mut socket: TcpStream, client_socket: Arc<Mutex<TcpStream>>, handler: Handler<MockSet<TestGossipableType>>) {
+
+    // Initialize a buffer of size 1024.
+    let mut buf = [0u8; 1024];
+
+    loop {
+        let n = socket.read(&mut buf).await.unwrap();
+
+        // Empty, wait 5 sec before next attempt
+        if n == 0 {
```

Comment:
When socket.read returns 0, that means the client has disconnected. You will forever get 0, so this is an infinite loop.

Reply:
I think if the client disconnects, the `read` call will error and the program will panic on that `unwrap`. If wouldn't infinite loop.

Reply:
That's not true. Example program: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d809a48e2983d4fe94833852a8a28961

Client drops the `tcp_stream` on line 35, and waits for the server to finish up before exiting. While it's in this state, read() repeatedly returns `Ok(0)`